File: Labs/Lab6.py
import unittest

def optimal_bst_cost(keys, freq):
    n = len(keys)
    dp = [[0] * n for _ in range(n)]
    
    for i in range(n):
        dp[i][i] = freq[i]  
    
    for length in range(2, n + 1):  
        for i in range(n - length + 1):
            j = i + length - 1
            dp[i][j] = float('inf')
            sum_freq = sum(freq[i:j + 1])
            
            for r in range(i, j + 1):
                cost = (dp[i][r - 1] if r > i else 0) + \
                (dp[r + 1][j] if r < j else 0) + sum_freq
                dp[i][j] = min(dp[i][j], cost)
    
    return dp[0][n - 1]

class TestOptimalBSTCost(unittest.TestCase):
    def test_example(self):
        keys = [5, 6]
        freq = [17, 25]
        self.assertEqual(optimal_bst_cost(keys, freq), 59)

    def test_single_key(self):
        keys = [10]
        freq = [5]
        self.assertEqual(optimal_bst_cost(keys, freq), 5)

if __name__ == "__main__":
    unittest.main()

File: Labs/Lab6.cpp
#include <iostream>
#include <vector>
#include <cassert>

int optimalBSTCost(const std::vector<int>& keys, const std::vector<int>& freq) {
    int n = keys.size();
    std::vector<std::vector<int>> dp(n, std::vector<int>(n, 0));

    for (int i = 0; i < n; ++i) {
        dp[i][i] = freq[i];
    }

    for (int len = 2; len <= n; ++len) {
        for (int i = 0; i <= n - len; ++i) {
            int j = i + len - 1;
            dp[i][j] = INT_MAX;
            int totalFreq = 0;

            for (int k = i; k <= j; ++k) {
                totalFreq += freq[k];
            }

            for (int r = i; r <= j; ++r) {
                int cost = (r > i ? dp[i][r - 1] : 0) + (r < j ? dp[r + 1][j] : 0) + totalFreq;
                dp[i][j] = std::min(dp[i][j], cost);
            }
        }
    }

    return dp[0][n - 1];
}

// Catch2 Unit Test
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("Optimal BST Cost") {
    REQUIRE(optimalBSTCost({5, 6}, {17, 25}) == 59);
    REQUIRE(optimalBSTCost({10}, {5}) == 5);
    REQUIRE(optimalBSTCost({1, 2, 3}, {10, 20, 30}) == 110);
}
